### PR TITLE
Core: Nested modules

### LINKED ISSUE #543: Allow nested suites (modules)?
A personal sore point for me when using QUnit is the lack of nested suites (`module`s) that are offered by other popular frameworks like Mocha and Jasmine.

I've given it some cursory thought and, if we agree it adds value, I don't foresee any major roadblocks to implementing that capability.  It would give us some options for richer reporting outputs as well, e.g. Mocha's sleek nested HTML report:

screen shot 2014-03-12 at 1 00 06 pm

### PR DESCRIPTION
**Pull request by Askelkana**
Allows nesting of modules by enabling passing function as third argument
to QUnit.module method.

Fixes #543

### CODE REVIEW

**Comment by Askelkana** [comment]
I signed the CLA prior to creating the pull request (losing my virginity with this). Don't know what happens after that...

**Comment by arthurvr** [comment]
@Askelkana You used your username when signing the CLA, not your full name.

**Comment by Askelkana** [comment]
Actually, I didn't. But I'll have another go...

**Comment by arthurvr** [comment]
@Askelkana Now the CLA appears to be signed with your full name, thanks. You'll have to update your git config to include your full name as well. Normally it should be fine then.

**Comment by Askelkana** [comment]
By way of interest, the CLA page reads as follows:

> The email address used to sign the CLA must match the email address in your git config. The author information for each commit in your Contributions will be cross-checked with this signature.

It does not mention that the name given is significant.

**Comment by Askelkana** [comment]
Ok, git config updated and commited. I assume the CI build fail is not critical, since there are several errors being thrown which have nothing to do with my changes.

**Comment by leobalter** [comment]
I'll review this tomorrow morning but you also need to run grunt in order to lint and check the code style.

This PR looks interesting, I'm looking forward to see it passing the on CI.

**Comment by leobalter** [comment]
current tests are working, but I also made this following test in your branch:

```js
QUnit.module( "multiple hooks", {
	beforeEach: function( assert ) {
		assert.ok( true, "root module beforeEach" );
	},
	afterEach: function( assert ) {
		assert.ok( true, "root module afterEach" );
	}
}, function() {
	QUnit.module( "nested module", {
		beforeEach: function( assert ) {
			assert.ok( true, "nested module beforeEach" );
		},
		afterEach: function( assert ) {
			assert.ok( true, "nested module afterEach" );
		}
	});

	QUnit.test( "all hooks must be called", function( assert ) {
		assert.expect( 4 );
	});
});
```

It failed as it only triggered the nested module hooks. This test should pass triggering all the chained modules' hooks.

IMHO, the only real advantage on having nested modules is to stack the hooks. Otherwise it would just create a fancy report.

cc @JamesMGreene.

**Comment by Askelkana** [comment]
I think your tests are incorrect: surely the `afterEach` would be called after the test has run, and therefore any assertions fired there will not be caught by the `expect` method. However, I have integrated the intent of the check into my original tests.

I wanted to wait for feedback before attempting this, even though it was mentioned in the original issue's comments. So thanks for the confirmation.

I would add that another advantage to nesting modules is that it clears up one's testing code no end, since it allows for much better arrangement of the tests.

I'll push a commit through soon.

**Comment by leobalter** [comment]
the expect also counts the assertions in the following afterEach hook.

Some other issues I noticed:
- the nested modules are not listed on the HTML Reporter
- selecting the root modules on the HTML Reporter filter doesn't run the tests on its respective nested modules

**Comment by Askelkana** [comment]
You are right. Tests still holding.

As for the HTMLReporter, at present the modules are not highlighted in any way. I didn't wish to presume how this should be handled, so I plumbed for simply showing the module path on the test row, as before.

I agree that ideally, some nesting mechanism should be portrayed by the reporter.

File: src/core.js [hunk 1]
```diff
@@ -67,8 +68,8 @@ var QUnit,
 	objectValues = function( obj ) {
 		var key, val,
 			vals = QUnit.is( "array", obj ) ? [] : {};
-		for ( key in obj ) {
-			if ( hasOwn.call( obj, key ) ) {
+		for (key in obj) {
+			if (hasOwn.call( obj, key )) {
```
Comment (leobalter) [inline src/core.js]: something is breaking the code style, and it's becoming hard to check the real diffs here.

Comment (Askelkana) [inline src/core.js]: Ok, I'll check it out. Everything passed the jshint and jscs tests, so I didn't notice these.

The actual changes this time are in the test.run and test.hook methods.

**Comment by Askelkana** [comment]
> Some other issues I noticed:
> 
> the nested modules are not listed on the HTML Reporter
> selecting the root modules on the HTML Reporter filter doesn't run the tests on its respective nested modules

Well, that may be so, but currently QUnit doesn't provide any support for modules in the HTML Reporter. The features you highlight here are really independent of my proposed changes, in that they could (and should) already be there.

**Comment by Askelkana** [comment]
Ok, I just pushed an addition that all tests belonging to a super-module are run when selecting the said module from the filter. Unless you have any other issues, I think this is good to go.

File: src/core.js [hunk 2]
```diff
@@ -2,14 +2,15 @@ var QUnit,
 	config,
 	onErrorFnPrev,
 	loggingCallbacks = {},
-	fileName = ( sourceFromStacktrace( 0 ) || "" ).replace( /(:\d+)+\)?/, "" ).replace( /.+\//, "" ),
+	fileName = ( sourceFromStacktrace( 0 ) || "" ).replace( /(:\d+)+\)?/, "" ).replace( /.+\//,
```
Comment (jzaefferer) [inline src/core.js]: There are lots of whitespace or linebreak changes like this. They don't belong in this PR, most of them seem just plain wrong. Can you address that? We can then do a proper review (more timely, hopefully).

File: src/test.js [hunk 3]
```diff
@@ -19,15 +19,16 @@ function Test( settings ) {
 
 	this.testId = generateHash( this.module.name, this.testName );
 
-	this.module.tests.push({
+	this.module.tests.push( {
 		name: this.testName,
 		testId: this.testId
-	});
+	} );
```
Comment (jzaefferer) [inline src/test.js]: These changes are fine, since they actually match the styleguide

File: src/test.js [hunk 4]
```diff
@@ -42,13 +43,13 @@ Test.prototype = {
 		if (
 
 			// Emit moduleStart when we're switching from one module to another
-			this.module !== config.previousModule ||
+		this.module !== config.previousModule ||
```
Comment (jzaefferer) [inline src/test.js]: This isn't, it just makes things worse.

File: src/test.js (hunk 3 above)
Comment (jzaefferer) [inline src/test.js]: But they should happen in a separate PR, not in one that introduces a new feature.

**Comment by Askelkana** [comment]
Apologies for the delay. Slowed by poor health and a garden to get ready for planting...

Anyway, I've fixed (or think I have) the style issues. I take your point about not trying to solve more than one problem at a time. But the code really could stand some cleaning: it's quite inconsistent in many cases, and certainly doesn't pass the jslint test.

**Comment by Askelkana** [comment]
Yes, all styles issues have now been resolved. Happy reviewing!

**Comment by jzaefferer** [comment]
@leobalter what do you think?

**Comment by leobalter** [comment]
I believe fixing the conflict issues won't be too hard. 

I'll merge this and then I'll propose an extra PR to make the QUnit.module functions to come with the `test` and `QUnit` as the arguments.

---

The `test` argument is just to have it as a local reference (avoiding globals without using namespaces, aka `QUnit.test`); The `QUnit` argument is to allow access to additional features, as `QUnit.skip`, `QUnit.config`, the proposed `QUnit.only`, etc.

A second following PR proposal is to also alias QUnit.module to the QUnit. This is more insane, but would allow a simple and dynamic use with these nested modules:

```js
import QUnit from 'qunitjs';

QUnit( "my first batch", ( test ) => {
  test( "test this", ( t ) => {
    t.ok( true, "it pass" );
  } );
} );
```

I'll open the respective issues to discuss them as soon as I merge this PR.

**Comment by gibson042** [comment]
This PR exhibits some potentially surprising behaviors:
* executing all of a module's own tests before entering its submodules (regardless of test/submodule definition order)
* always minting a new environment definition for every submodule, even if its `hooks` argument is `undefined` or not even specified

If they are intentional, there should be assertions covering them. If they are unintentional and undesired, there should be assertions covering the desired behavior.

**Comment by leobalter** [comment]
The squashed and rebased branch is here: https://github.com/jquery/qunit/compare/jquery:master...leobalter:nested-modules?expand=1

---

@gibson042 

> executing all of a module's own tests before entering its submodules (regardless of test/submodule definition order)

This might seem weird, but with the idea of each test should be atomic, an specific order shouldn't be a major concern. The real world prevents me to keep this argument as this will indeed blow out eventually without assertions with the intentional behavior.

> always minting a new environment definition for every submodule, even if its hooks argument is undefined or not even specified

I don't see it clearly, but I believe this and along with the other behavior could be improved in a next PR?

The main feature seems ok and working fine, and I am late giving the proper attention to this work. So if you agree, I would let these parts to be treated in a next commit bit.

**Comment by gibson042** [comment]
> The main feature seems ok and working fine, and I am late giving the proper attention to this work. So if you agree, I would let these parts to be treated in a next commit bit.

:+1:

I just didn't want to lose these thoughts (or the one regarding [explicit context sensitivity of `beforeEach`/`afterEach`/`test`/etc.](https://github.com/jquery/qunit/pull/670#discussion_r26321413) via arguments or context, which should be covered by your above followup).

File: test/modules.js [hunk 5]
```diff
@@ -147,3 +147,51 @@ QUnit.module( "Deprecated setup/teardown", {
...
+				assert.expect( 3 );
+			});
+		});
+	QUnit.module( "second outer" );
+	QUnit.test( "test after all nesting modules processed and new module declared",
+		function( assert ) {
+			var module = assert.test.module;
+			assert.equal( module.name, "nested modules > second outer" );
+		});
+});
```
Comment (leobalter) [inline test/modules.js]: I found an interesting and possibly blocking behavior:

`QUnit.module` calls with functions are not affecting the following tests written outside the function.

Like this following example right after this line where I am commenting:

```js
QUnit.test( "test foo", function( a ) {
	a.equal( a.test.module.name, "nested modules" );
} );
```

This test is failing, telling the actual value is `"Deprecated setup/teardown"`, referencing the module written above `"nested modules"`.

Comment (leobalter) [inline test/modules.js]: @jzaefferer The implementation is tied to this behavior.

I believe this can still be shipped but mentioning that calling `QUnit.module` with a nested function for tests will not spread the current module for consequent `QUnit.test` our of its nested function.

This should be tested, documented and we might want to deprecate the usage of `QUnit.module` without a function parameter.

Comment (leobalter) [inline test/modules.js]: also: calling `QUnit.module` without a function inside a `QUnit.module` function will cause a strong mess with the nested behavior. That's a reason it should be deprecated.

Comment (leobalter) [inline test/modules.js]: please, see #859

**Comment by leobalter** [comment]
Closing this, further work in progress on #859
